### PR TITLE
Update Latebound Add Collection Initializer handling

### PR DESCRIPTION
**Pull request by 333fred**
Adjusted assert and lateboundaccesskind for latebound
Add method on a collection initializer. Additionally
adjusted compiler-generated detection and added a test.

Fixes https://github.com/dotnet/roslyn/issues/27034.

@dotnet/roslyn-compiler @AlekseyTs @dotnet/analyzer-ioperation for review.

### CODE REVIEW

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.vb
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.vb]: Ensures that the member reference in the explicit add call is actually explicit in the tree.

**Comment by etbyrd** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by etbyrd** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Latebound.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Latebound.vb]: >eceiver.Syntax.Kind = SyntaxKind.ObjectCreationExpression [](start = 56, length = 57)

Can we use something other than a syntax check? Are we using special bound nodes as receivers in this case?

File: src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb]: >( [](start = 19, length = 1)

Parens are not neede in VB. Consider removing, given that the condition is very simple

File: src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb]: >invocation = DirectCast(invocation, BoundLateInvocation).SetLateBoundAccessKind(LateBoundAccessKind.Call) [](start = 20, length = 105)

Could you please confirm (by examining IL) that we keep calling the same VB runtime helper after this change and also compare against native compiler?

File: src/Compilers/VisualBasic/Test/Semantic/Binding/BindingCollectionInitializerTests.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Binding/BindingCollectionInitializerTests.vb]: ><CDATA[ [

Could probably be just a string literal, since this is just a single line

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

**Comment by etbyrd** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by etbyrd** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Latebound.vb
Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Binding/Binder_Latebound.vb]: Worked with @AlekseyTs offline to find a better approach.

---
In reply to: [190969490](https://github.com/dotnet/roslyn/pull/27126#discussion_r190969490) [](ancestors = 190969490)

File: src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb
Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb]: This change brings us inline with the native compiler, looks like we had a bug here. Previous to this change, we were calling `[Microsoft.VisualBasic]Microsoft.VisualBasic.CompilerServices.NewLateBinding::LateGet`, whereas the native compiler called: `[Microsoft.VisualBasic]Microsoft.VisualBasic.CompilerServices.NewLateBinding::LateCall`. With this change, we also call `LateCall`.

---
In reply to: [190973564](https://github.com/dotnet/roslyn/pull/27126#discussion_r190973564) [](ancestors = 190973564)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [392144689](https://github.com/dotnet/roslyn/pull/27126#issuecomment-392144689) [](ancestors = 392144689)

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3). Please follow up with Compatibility Council and see if we can completely remove WasCompilerGenerated fixup in BindLateBoundMemberAccess (can be done in a separate PR).

**Review by mavasani — APPROVED** [review]

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler need a second review please.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @jaredpar @cston need a second review please

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.vb
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.vb]: Minor point: `Mod1` seems unnecessary, here and in other `IOperation` test.

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@333fred, @cston, @dotnet/roslyn-compiler Please review

**Pull request merged**
